src/Polly.Core/Utils/CancellationTokenSourcePool.cs

```diff
@@ -1,33 +1,78 @@
+using System.Diagnostics.CodeAnalysis;
+
 namespace Polly.Utils;
 
 #pragma warning disable CA1716 // Identifiers should not match keywords
 
-internal abstract partial class CancellationTokenSourcePool
+internal sealed class CancellationTokenSourcePool
 {
+#if NET6_0_OR_GREATER
+    private static readonly CancellationTokenSourcePool SystemInstance = new(TimeProvider.System);
+
+    private readonly ObjectPool<CancellationTokenSource> _pool = new(
+        static () => new CancellationTokenSource(),
+        static cts => true);
+#endif
+
+    private readonly TimeProvider _timeProvider;
+
     public static CancellationTokenSourcePool Create(TimeProvider timeProvider)
     {
 #if NET6_0_OR_GREATER
         if (timeProvider == TimeProvider.System)
         {
-            return PooledCancellationTokenSourcePool.SystemInstance;
+            return SystemInstance;
         }
 #endif
-        return new DisposableCancellationTokenSourcePool(timeProvider);
+
+        return new(timeProvider);
     }
 
+    public CancellationTokenSourcePool(TimeProvider timeProvider) => _timeProvider = timeProvider;
+
     public CancellationTokenSource Get(TimeSpan delay)
     {
         if (delay <= TimeSpan.Zero && delay != System.Threading.Timeout.InfiniteTimeSpan)
         {
             throw new ArgumentOutOfRangeException(nameof(delay), delay, "Invalid delay specified.");
         }
 
-        return GetCore(delay);
+        CancellationTokenSource source;
+
+#if NET6_0_OR_GREATER
+        if (this == SystemInstance)
+        {
+            source = _pool.Get();
+        }
+        else
+        {
+            source = new CancellationTokenSource();
+        }
+#else
+        source = new CancellationTokenSource();
+#endif
+
+        if (IsCancellable(delay))
+        {
+            _timeProvider.CancelAfter(source, delay);
+        }
+
+        return source;
     }
 
-    protected abstract CancellationTokenSource GetCore(TimeSpan delay);
+    [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Can't be static on >= .NET 6")]
+    public void Return(CancellationTokenSource source)
+    {
+#if NET6_0_OR_GREATER
+        if (this == SystemInstance && source.TryReset())
```

Comment:
nit: Isn't now this extra if check that's always executed worse than the method call?

Reply:
Yes, there is an extra "if" check when on .NET 6 or above. But you trade this for eliminating virtual dispatch and enabling inlining.